File: aoc_2022/src/main/java/me/frenz/day13/PacketParser.java
package me.frenz.day13;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

class PacketParser {

    private final Pattern number = Pattern.compile("^(\\d+)\\D*");

    String parsePacket(PacketValue packet, String line) {
        while (!line.isBlank()) {
            if (line.startsWith("[")) {
                ContainerPacket childs = new ContainerPacket();
                line = parsePacket(childs, line.substring(1));
                packet.add(childs);
            }
            if (line.startsWith("]")) {
                return line.substring(1);
            }
            final Matcher m = number.matcher(line);

            if (m.find()) {
                int num = Integer.parseInt(m.group(1));
                packet.add(new IntPacket(num));
                line = line.substring(m.group(1).length());
            }
            if (line.startsWith(",")) {
                line = line.substring(1);
            }
        }

        return null;
    }
}

File: aoc_2022/src/main/java/me/frenz/day13/PacketParser.kt
package me.frenz.day13

import java.util.*

internal class PacketParser {

    fun parsePacket(packet: String): PacketValue {
        val stack = Stack<PacketValue>()
        stack.push(ContainerPacket())

        var dupPacket = packet
        while (dupPacket.isNotEmpty()) {
            when (dupPacket[0]) {
                '[' -> {
                    stack.push(ContainerPacket())
                    dupPacket = dupPacket.substring(1)
                }
                ']' -> {
                    val topItem = stack.pop()
                    (stack.peek() as ContainerPacket).add(topItem)
                    dupPacket = dupPacket.substring(1)
                }
                ',' -> {
                    dupPacket = dupPacket.substring(1)
                }
                else -> {
                    // it's a digit
                    val digString = StringBuilder()
                    var i = 0
                    while (dupPacket[i].isDigit()) {
                        digString.append(dupPacket[i++])
                    }
                    (stack.peek() as ContainerPacket).add(IntPacket(digString.toString().toInt()))
                    dupPacket = dupPacket.substring(digString.length)
                }
            }
        }
        return (stack.pop() as ContainerPacket).unwrap()
    }
}
